plugins/github-release-manager/src/api/PluginApiClient.ts

```diff
@@ -0,0 +1,961 @@
+
+type Merge = ({
+  base,
+  head,
+}: {
+  base: string;
+  head: string;
+} & OwnerRepo) => Promise<{
+  htmlUrl: string;
+  commit: {
+    message: string;
+    tree: {
+      sha: string;
+    };
+  };
+}>;
+export type MergeResult = UnboxReturnedPromise<Merge>;
+
+type CreateCherryPickCommit = (
+  args: {
+    bumpedTag: string;
+    selectedPatchCommit: UnboxArray<
+      UnboxReturnedPromise<IPluginApiClient['getRecentCommits']>
+    >;
+    mergeTree: string;
+    releaseBranchSha: string;
+  } & OwnerRepo,
+) => Promise<{
+  message: string;
+  sha: string;
+}>;
+export type CreateCherryPickCommitResult = UnboxReturnedPromise<CreateCherryPickCommit>;
+
+type ReplaceTempCommit = (
+  args: {
+    releaseBranchName: string;
+    cherryPickCommit: UnboxReturnedPromise<
+      IPluginApiClient['patch']['createCherryPickCommit']
+    >;
+  } & OwnerRepo,
+) => Promise<{
+  ref: string;
+  object: {
+    sha: string;
+  };
+}>;
+export type ReplaceTempCommitResult = UnboxReturnedPromise<ReplaceTempCommit>;
+
+type CreateTagObject = ({
+  bumpedTag,
+  updatedReference,
+}: {
+  bumpedTag: string;
+  updatedReference: ReplaceTempCommitResult;
+} & OwnerRepo) => Promise<{
+  tag: string;
+  sha: string;
+}>;
+export type CreateTagObjectResult = UnboxReturnedPromise<CreateTagObject>;
+
+type CreateReference = (
+  args: {
+    bumpedTag: string;
+    createdTagObject: CreateTagObjectResult;
+  } & OwnerRepo,
+) => Promise<{
+  ref: string;
+}>;
+export type CreateReferenceResult = UnboxReturnedPromise<CreateReference>;
+
+type UpdateRelease = (
+  args: {
+    bumpedTag: string;
+    latestRelease: NonNullable<GetLatestReleaseResult>;
+    tagParts: SemverTagParts | CalverTagParts;
+    selectedPatchCommit: GetRecentCommitsResultSingle;
+  } & OwnerRepo,
+) => Promise<{
+  name: string | null;
+  tagName: string;
+  htmlUrl: string;
+}>;
+export type UpdateReleaseResult = UnboxReturnedPromise<UpdateRelease>;
+
+/**
+ * PromoteRc
+ */
+type PromoteRelease = (
+  args: {
+    releaseId: NonNullable<GetLatestReleaseResult>['id'];
+    releaseVersion: string;
+  } & OwnerRepo,
+) => Promise<{
+  name: string | null;
+  tagName: string;
+  htmlUrl: string;
+}>;
+export type PromoteReleaseResult = UnboxReturnedPromise<PromoteRelease>;
+
+/**
+ * Stats
+ */
+type GetAllTags = (
+  args: OwnerRepo,
+) => Promise<
+  Array<{
+    tagName: string;
+    sha: string;
+  }>
+>;
+export type GetAllTagsResult = UnboxReturnedPromise<GetAllTags>;
+
+type GetAllReleases = (
+  args: OwnerRepo,
+) => Promise<
+  Array<{
+    id: number;
+    name: string | null;
+    tagName: string;
+    createdAt: string | null;
+    htmlUrl: string;
+  }>
+>;
+export type GetAllReleasesResult = UnboxReturnedPromise<GetAllReleases>;
+
+type GetCommit = (
+  args: {
+    ref: string;
+  } & OwnerRepo,
+) => Promise<{
+  createdAt: string | undefined;
+}>;
+export type GetCommitResult = UnboxReturnedPromise<GetCommit>;
+
+export interface IPluginApiClient {
```

Comment:
We generally avoid the `I` prefix and go with a `Api` suffix instead. I'm assuming "plugin" here refers to this plugin? In that case I think a good name could be `Git(hub)Release(Management)Api`.

This API is also incredibly broad and low-level, I'm wondering if it's possible to lift it up a level or two? It would involve lifting logic out from e.g. the hooks and into this API. Just something I smell at surface level though, haven't had a deep look at the design to see if it makes sense. Also something that could be worked on in followup PRs

Reply:
> 

We generally avoid the I prefix and go with a Api suffix instead. I'm assuming "plugin" here refers to this plugin? In that case I think a good name could be Git(hub)Release(Management)Api.

👍 

> This API is also incredibly broad and low-level, I'm wondering if it's possible to lift it up a level or two? It would involve lifting logic out from e.g. the hooks and into this API. Just something I smell at surface level though, haven't had a deep look at the design to see if it makes sense. Also something that could be worked on in followup PRs

It is indeed low-level. The intention is to spare the potential user building an override the hassle of figuring out how various git commands relate to one another (there's a bit of work left with refining the API method's signatures to make them more agnostic).

Furthermore, by having low level methods I'm able to provide gradual UI updates with the progress and displaying what's going on. Moving entire an entire feature's requests to a single API method would remove this ability.